File: frontend/src/lib/api/content.ts
/**
 * API Content — mappée sur les endpoints backend /api/v1/content/*
 * Gère: News, FAQ, Editions, Éditions passées, Partenaires, Pages
 * La conversion snake_case ↔ camelCase est automatique via les intercepteurs du client
 */
import apiClient from './client'

// ==================== NEWS / ACTUALITÉS ====================

export interface NewsItem {
    id: string
    title: string
    content: string
    excerpt?: string
    imageUrl?: string
    author?: string
    category?: string
    isPublished: boolean
    publishedAt?: string
    externalUrl?: string  // Lien vers l'article original
    createdAt: string
    updatedAt: string
}

export interface NewsCreate {
    title: string
    content: string
    excerpt?: string
    imageUrl?: string
    category?: string
    isPublished?: boolean
    externalUrl?: string  // Lien vers l'article original
}

export interface NewsUpdate {
    title?: string
    content?: string
    excerpt?: string
    imageUrl?: string
    category?: string
    isPublished?: boolean
    externalUrl?: string  // Lien vers l'article original
}

// ==================== FAQ ====================

export interface FAQItem {
    id: string
    question: string
    answer: string
    category?: string
    order: number
    isPublished: boolean
    createdAt: string
}

export interface FAQCreate {
    question: string
    answer: string
    category?: string
    order?: number
    isPublished?: boolean
}

export interface FAQUpdate {
    question?: string
    answer?: string
    category?: string
    order?: number
    isPublished?: boolean
}

// ==================== EDITIONS ====================

export interface TimelinePhase {
    id: string
    phaseOrder: number
    title: string
    description?: string
    startDate?: string
    endDate?: string
    isCurrent: boolean
}

export interface TimelinePhaseCreate {
    phaseOrder: number
    title: string
    description?: string
    startDate?: string
    endDate?: string
    isCurrent?: boolean
}

export interface CalendarEvent {
    id: string
    title: string
    description?: string
    eventDate: string
    eventType?: string
}

export interface SelectionCriterion {
    id: string
    stage: string
    stageOrder: number
    criterion: string
    minScore?: number
}

export interface EditionPartner {
    id: string
    partnerId: string
    role?: string
    contribution?: string
}

export interface Edition {
    id: string
    year: number
    isActive: boolean
    title: string
    description?: string
    timelinePhases: TimelinePhase[]
    calendarEvents: CalendarEvent[]
    selectionCriteria: SelectionCriterion[]
    editionPartners: EditionPartner[]
    createdAt: string
}

export interface EditionCreate {
    year: number
    isActive?: boolean
    title: string
    description?: string
}

export interface EditionUpdate {
    year?: number
    isActive?: boolean
    title?: string
    description?: string
}

// ==================== ÉDITIONS PASSÉES ====================

export interface PastTimelinePhase {
    id: string
    phaseOrder: number
    title: string
    description?: string
    date?: string
}

export interface GalleryImage {
    id: string
    imageUrl: string
    caption?: string
    order: number
}

export interface Testimonial {
    id: string
    studentName: string
    school?: string
    role?: string
    quote: string
    imageUrl?: string
    pastEditionId?: string
}

// ==================== GENERAL TESTIMONIALS ====================

export interface GeneralTestimonial {
    id: string
    authorName: string
    authorRole?: string
    authorType?: string  // "mentor", "parent", "sponsor", "partner"
    content: string
    photoUrl?: string
    videoUrl?: string
    organization?: string
    displayOrder?: number
    isPublished: boolean
}

export interface GeneralTestimonialCreate {
    authorName: string
    authorRole?: string
    authorType?: string
    content: string
    photoUrl?: string
    videoUrl?: string
    organization?: string
    displayOrder?: number
    isPublished?: boolean
}

export interface GeneralTestimonialUpdate {
    authorName?: string
    authorRole?: string
    authorType?: string
    content?: string
    photoUrl?: string
    videoUrl?: string
    organization?: string
    displayOrder?: number
    isPublished?: boolean
}

export interface Achievement {
    id: string
    title: string
    description?: string
    category?: string
    rank?: number
}

export interface PressRelease {
    id: string
    title: string
    source?: string
    url?: string
    publishedAt?: string
}

export interface EditionStat {
    id: string
    metricName: string
    metricValue: number
    metricUnit?: string
}

export interface PastEdition {
    id: string
    year: number
    hostCountry?: string
    numStudents?: number
    pastTimelinePhases: PastTimelinePhase[]
    galleryImages: GalleryImage[]
    testimonials: Testimonial[]
    achievements: Achievement[]
    pressReleases: PressRelease[]
    editionStats: EditionStat[]
    createdAt: string
}

export interface PastEditionCreate {
    year: number
    hostCountry?: string
    numStudents?: number
}

export interface PastEditionUpdate {
    year?: number
    hostCountry?: string
    numStudents?: number
}

// ==================== PARTENAIRES ====================

export interface Partner {
    id: string
    name: string
    logoUrl?: string
    description?: string
    websiteUrl?: string
    order: number
    isActive: boolean
}

export interface PartnerCreate {
    name: string
    logoUrl?: string
    description?: string
    websiteUrl?: string
    order?: number
    isActive?: boolean
}

export interface PartnerUpdate {
    name?: string
    logoUrl?: string
    description?: string
    websiteUrl?: string
    order?: number
    isActive?: boolean
}

// ==================== PAGES ====================

export interface Page {
    id: string
    slug: string
    title: string
    content?: string
    metaData?: Record<string, any>
    isPublished: boolean
    createdAt: string
}

export interface PageCreate {
    slug: string
    title: string
    content?: string
    metaData?: Record<string, any>
    isPublished?: boolean
}

export interface PageUpdate {
    slug?: string
    title?: string
    content?: string
    metaData?: Record<string, any>
    isPublished?: boolean
}

// ==================== NEXT DEADLINE ====================

export interface NextDeadline {
    phaseTitle: string
    phaseDescription?: string
    targetDate: string
    targetType: 'start' | 'end'
    currentPhase?: {
        title: string
        isActive: boolean
        startDate?: string
        endDate?: string
    }
    editionYear: number
}

// ==================== PUBLIC STATS ====================

export interface PublicStats {
    totalCandidates: number
    verifiedCandidates: number
}

// ==================== API CALLS ====================

export const contentApi = {
    // ==================== NEWS ====================

    /**
     * Liste des actualités
     * GET /content/news
     */
    getNews: async (params?: {
        skip?: number
        limit?: number
        publishedOnly?: boolean
    }): Promise<NewsItem[]> => {
        const { data } = await apiClient.get('/content/news', { params })
        return data
    },

    /**
     * Détails d'une actualité
     * GET /content/news/:newsId
     */
    getNewsItem: async (newsId: string): Promise<NewsItem> => {
        const { data } = await apiClient.get(`/content/news/${newsId}`)
        return data
    },

    /**
     * Créer une actualité (Admin)
     * POST /content/news
     */
    createNews: async (newsData: NewsCreate): Promise<NewsItem> => {
        const { data } = await apiClient.post('/content/news', newsData)
        return data
    },

    /**
     * Mettre à jour une actualité (Admin)
     * PUT /content/news/:newsId
     */
    updateNews: async (newsId: string, newsData: NewsUpdate): Promise<NewsItem> => {
        const { data } = await apiClient.put(`/content/news/${newsId}`, newsData)
        return data
    },

    /**
     * Supprimer une actualité (Admin)
     * DELETE /content/news/:newsId
     */
    deleteNews: async (newsId: string): Promise<void> => {
        await apiClient.delete(`/content/news/${newsId}`)
    },

    // ==================== FAQ ====================

    /**
     * Liste des FAQ
     * GET /content/faq
     */
    getFAQ: async (params?: {
        category?: string
        publishedOnly?: boolean
    }): Promise<FAQItem[]> => {
        const { data } = await apiClient.get('/content/faq', { params })
        return data
    },

    /**
     * Créer une FAQ (Admin)
     * POST /content/faq
     */
    createFAQ: async (faqData: FAQCreate): Promise<FAQItem> => {
        const { data } = await apiClient.post('/content/faq', faqData)
        return data
    },

    /**
     * Mettre à jour une FAQ (Admin)
     * PUT /content/faq/:faqId
     */
    updateFAQ: async (faqId: string, faqData: FAQUpdate): Promise<FAQItem> => {
        const { data } = await apiClient.put(`/content/faq/${faqId}`, faqData)
        return data
    },

    /**
     * Supprimer une FAQ (Admin)
     * DELETE /content/faq/:faqId
     */
    deleteFAQ: async (faqId: string): Promise<void> => {
        await apiClient.delete(`/content/faq/${faqId}`)
    },

    // ==================== EDITIONS ====================

    /**
     * Récupérer l'édition active
     * GET /content/editions/active
     */
    getActiveEdition: async (): Promise<Edition | null> => {
        const { data } = await apiClient.get('/content/editions/active')
        return data
    },

    /**
     * Récupérer la prochaine deadline pour le compte à rebours
     * GET /content/editions/active/next-deadline
     */
    getNextDeadline: async (): Promise<NextDeadline | null> => {
        const { data } = await apiClient.get('/content/editions/active/next-deadline')
        return data
    },

    /**
     * Liste de toutes les éditions
     * GET /content/editions
     */
    getEditions: async (params?: {
        skip?: number
        limit?: number
    }): Promise<Edition[]> => {
        const { data } = await apiClient.get('/content/editions', { params })
        return data
    },

    /**
     * Détails d'une édition
     * GET /content/editions/:editionId
     */
    getEdition: async (editionId: string): Promise<Edition> => {
        const { data } = await apiClient.get(`/content/editions/${editionId}`)
        return data
    },

    /**
     * Créer une édition (Admin)
     * POST /content/editions
     */
    createEdition: async (editionData: EditionCreate): Promise<Edition> => {
        const { data } = await apiClient.post('/content/editions', editionData)
        return data
    },

    /**
     * Mettre à jour une édition (Admin)
     * PUT /content/editions/:editionId
     */
    updateEdition: async (editionId: string, editionData: EditionUpdate): Promise<Edition> => {
        const { data } = await apiClient.put(`/content/editions/${editionId}`, editionData)
        return data
    },

    // ==================== PHASES D'ÉDITION ====================

    /**
     * Créer une phase pour une édition (Admin)
     * POST /content/editions/:editionId/phases
     */
    createEditionPhase: async (editionId: string, phaseData: TimelinePhaseCreate): Promise<TimelinePhase> => {
        const { data } = await apiClient.post(`/content/editions/${editionId}/phases`, phaseData)
        return data
    },

    /**
     * Mettre à jour une phase (Admin)
     * PUT /content/editions/:editionId/phases/:phaseId
     */
    updateEditionPhase: async (editionId: string, phaseId: string, phaseData: TimelinePhaseCreate): Promise<TimelinePhase> => {
        const { data } = await apiClient.put(`/content/editions/${editionId}/phases/${phaseId}`, phaseData)
        return data
    },

    /**
     * Supprimer une phase (Admin)
     * DELETE /content/editions/:editionId/phases/:phaseId
     */
    deleteEditionPhase: async (editionId: string, phaseId: string): Promise<void> => {
        await apiClient.delete(`/content/editions/${editionId}/phases/${phaseId}`)
    },

    // ==================== CRITÈRES DE SÉLECTION ====================

    /**
     * Ajouter un critère de sélection (Admin)
     * POST /content/editions/:editionId/criteria
     */
    createEditionCriterion: async (editionId: string, criterionData: {
        stage: string
        stageOrder: number
        criterion: string
        minScore?: number
    }): Promise<SelectionCriterion> => {
        const { data } = await apiClient.post(`/content/editions/${editionId}/criteria`, criterionData)
        return data
    },

    /**
     * Mettre à jour un critère (Admin)
     * PUT /content/editions/:editionId/criteria/:criterionId
     */
    updateEditionCriterion: async (editionId: string, criterionId: string, criterionData: {
        stage: string
        stageOrder: number
        criterion: string
        minScore?: number
    }): Promise<SelectionCriterion> => {
        const { data } = await apiClient.put(`/content/editions/${editionId}/criteria/${criterionId}`, criterionData)
        return data
    },

    /**
     * Supprimer un critère (Admin)
     * DELETE /content/editions/:editionId/criteria/:criterionId
     */
    deleteEditionCriterion: async (editionId: string, criterionId: string): Promise<void> => {
        await apiClient.delete(`/content/editions/${editionId}/criteria/${criterionId}`)
    },

    // ==================== ÉDITIONS PASSÉES ====================

    /**
     * Liste des éditions passées
     * GET /content/past-editions
     */
    getPastEditions: async (): Promise<PastEdition[]> => {
        const { data } = await apiClient.get('/content/past-editions')
        return data
    },

    /**
     * Détails d'une édition passée
     * GET /content/past-editions/:editionId
     */
    getPastEdition: async (editionId: string): Promise<PastEdition> => {
        const { data } = await apiClient.get(`/content/past-editions/${editionId}`)
        return data
    },

    /**
     * Créer une édition passée (Admin)
     * POST /content/past-editions
     */
    createPastEdition: async (editionData: PastEditionCreate): Promise<PastEdition> => {
        const { data } = await apiClient.post('/content/past-editions', editionData)
        return data
    },

    /**
     * Mettre à jour une édition passée (Admin)
     * PUT /content/past-editions/:editionId
     */
    updatePastEdition: async (editionId: string, editionData: PastEditionUpdate): Promise<PastEdition> => {
        const { data } = await apiClient.put(`/content/past-editions/${editionId}`, editionData)
        return data
    },

    /**
     * Récupérer les témoignages de l'édition la plus récente
     * Affiche automatiquement les témoignages de la dernière édition sur la page d'accueil
     */
    /**
     * Récupérer tous les témoignages généraux pour la page d'accueil
     * Utilise les témoignages généraux (mentors, parents, sponsors, etc.)
     */
    getAllTestimonials: async (): Promise<GeneralTestimonial[]> => {
        return contentApi.getGeneralTestimonials(true)
    },

    // ==================== PARTENAIRES ====================

    /**
     * Liste des partenaires
     * GET /content/partners
     */
    getPartners: async (params?: {
        activeOnly?: boolean
    }): Promise<Partner[]> => {
        const { data } = await apiClient.get('/content/partners', { params })
        return data
    },

    /**
     * Créer un partenaire (Admin)
     * POST /content/partners
     */
    createPartner: async (partnerData: PartnerCreate): Promise<Partner> => {
        const { data } = await apiClient.post('/content/partners', partnerData)
        return data
    },

    /**
     * Mettre à jour un partenaire (Admin)
     * PUT /content/partners/:partnerId
     */
    updatePartner: async (partnerId: string, partnerData: PartnerUpdate): Promise<Partner> => {
        const { data } = await apiClient.put(`/content/partners/${partnerId}`, partnerData)
        return data
    },

    /**
     * Supprimer un partenaire (Admin)
     * DELETE /content/partners/:partnerId
     */
    deletePartner: async (partnerId: string): Promise<void> => {
        await apiClient.delete(`/content/partners/${partnerId}`)
    },

    // ==================== PAGES ====================

    /**
     * Récupérer une page par slug
     * GET /content/pages/:slug
     */
    getPageBySlug: async (slug: string): Promise<Page> => {
        const { data } = await apiClient.get(`/content/pages/${slug}`)
        return data
    },

    /**
     * Liste de toutes les pages
     * GET /content/pages
     */
    getPages: async (params?: {
        publishedOnly?: boolean
    }): Promise<Page[]> => {
        const { data } = await apiClient.get('/content/pages', { params })
        return data
    },

    /**
     * Créer une page (Admin)
     * POST /content/pages
     */
    createPage: async (pageData: PageCreate): Promise<Page> => {
        const { data } = await apiClient.post('/content/pages', pageData)
        return data
    },

    /**
     * Mettre à jour une page (Admin)
     * PUT /content/pages/:pageId
     */
    updatePage: async (pageId: string, pageData: PageUpdate): Promise<Page> => {
        const { data } = await apiClient.put(`/content/pages/${pageId}`, pageData)
        return data
    },

    /**
     * Supprimer une page (Admin)
     * DELETE /content/pages/:pageId
     */
    deletePage: async (pageId: string): Promise<void> => {
        await apiClient.delete(`/content/pages/${pageId}`)
    },

    // ==================== STATS PUBLIQUES ====================

    /**
     * Récupérer les statistiques publiques pour le site vitrine
     * GET /content/public-stats
     */
    getPublicStats: async (): Promise<PublicStats> => {
        const { data } = await apiClient.get('/content/public-stats')
        return data
    },

    // ==================== PAST EDITION TIMELINE PHASES ====================

    /**
     * Créer une phase de timeline pour une édition passée (Admin)
     * POST /content/past-editions/:editionId/timeline
     */
    createTimelinePhase: async (editionId: string, phaseData: {
        phaseOrder: number;
        title: string;
        description?: string;
        date?: string;
    }): Promise<PastTimelinePhase> => {
        const { data } = await apiClient.post(`/content/past-editions/${editionId}/timeline`, phaseData)
        return data
    },

    /**
     * Mettre à jour une phase de timeline (Admin)
     * PUT /content/past-editions/:editionId/timeline/:phaseId
     */
    updateTimelinePhase: async (editionId: string, phaseId: string, phaseData: {
        phaseOrder: number;
        title: string;
        description?: string;
        date?: string;
    }): Promise<PastTimelinePhase> => {
        const { data } = await apiClient.put(`/content/past-editions/${editionId}/timeline/${phaseId}`, phaseData)
        return data
    },

    /**
     * Supprimer une phase de timeline (Admin)
     * DELETE /content/past-editions/:editionId/timeline/:phaseId
     */
    deleteTimelinePhase: async (editionId: string, phaseId: string): Promise<void> => {
        await apiClient.delete(`/content/past-editions/${editionId}/timeline/${phaseId}`)
    },

    // ==================== PAST EDITION GALLERY IMAGES ====================

    /**
     * Créer une image de galerie pour une édition passée (Admin)
     * POST /content/past-editions/:editionId/gallery
     */
    createGalleryImage: async (editionId: string, imageData: {
        imageUrl: string;
        caption?: string;
        order: number;
    }): Promise<GalleryImage> => {
        const { data } = await apiClient.post(`/content/past-editions/${editionId}/gallery`, imageData)
        return data
    },

    /**
     * Mettre à jour une image de galerie (Admin)
     * PUT /content/past-editions/:editionId/gallery/:imageId
     */
    updateGalleryImage: async (editionId: string, imageId: string, imageData: {
        imageUrl: string;
        caption?: string;
        order: number;
    }): Promise<GalleryImage> => {
        const { data } = await apiClient.put(`/content/past-editions/${editionId}/gallery/${imageId}`, imageData)
        return data
    },

    /**
     * Supprimer une image de galerie (Admin)
     * DELETE /content/past-editions/:editionId/gallery/:imageId
     */
    deleteGalleryImage: async (editionId: string, imageId: string): Promise<void> => {
        await apiClient.delete(`/content/past-editions/${editionId}/gallery/${imageId}`)
    },

    // ==================== PAST EDITION TESTIMONIALS ====================

    /**
     * Créer un témoignage pour une édition passée (Admin)
     * POST /content/past-editions/:editionId/testimonials
     */
    createTestimonial: async (editionId: string, testimonialData: {
        studentName: string;
        school?: string;
        role?: string;
        quote: string;
        imageUrl?: string;
    }): Promise<Testimonial> => {
        const { data } = await apiClient.post(`/content/past-editions/${editionId}/testimonials`, testimonialData)
        return data
    },

    /**
     * Mettre à jour un témoignage (Admin)
     * PUT /content/past-editions/:editionId/testimonials/:testimonialId
     */
    updateTestimonial: async (editionId: string, testimonialId: string, testimonialData: {
        studentName: string;
        school?: string;
        role?: string;
        quote: string;
        imageUrl?: string;
    }): Promise<Testimonial> => {
        const { data } = await apiClient.put(`/content/past-editions/${editionId}/testimonials/${testimonialId}`, testimonialData)
        return data
    },

    /**
     * Supprimer un témoignage (Admin)
     * DELETE /content/past-editions/:editionId/testimonials/:testimonialId
     */
    deleteTestimonial: async (editionId: string, testimonialId: string): Promise<void> => {
        await apiClient.delete(`/content/past-editions/${editionId}/testimonials/${testimonialId}`)
    },

    // ==================== GENERAL TESTIMONIALS ====================

    /**
     * Récupérer tous les témoignages généraux (Public/Admin)
     * GET /content/general-testimonials
     */
    getGeneralTestimonials: async (publishedOnly: boolean = true): Promise<GeneralTestimonial[]> => {
        const { data } = await apiClient.get('/content/general-testimonials', {
            params: { published_only: publishedOnly }
        })
        return data
    },

    /**
     * Créer un témoignage général (Admin)
     * POST /content/general-testimonials
     */
    createGeneralTestimonial: async (testimonialData: GeneralTestimonialCreate): Promise<GeneralTestimonial> => {
        const { data } = await apiClient.post('/content/general-testimonials', testimonialData)
        return data
    },

    /**
     * Mettre à jour un témoignage général (Admin)
     * PUT /content/general-testimonials/:testimonialId
     */
    updateGeneralTestimonial: async (testimonialId: string, testimonialData: GeneralTestimonialUpdate): Promise<GeneralTestimonial> => {
        const { data } = await apiClient.put(`/content/general-testimonials/${testimonialId}`, testimonialData)
        return data
    },

    /**
     * Supprimer un témoignage général (Admin)
     * DELETE /content/general-testimonials/:testimonialId
     */
    deleteGeneralTestimonial: async (testimonialId: string): Promise<void> => {
        await apiClient.delete(`/content/general-testimonials/${testimonialId}`)
    },
}

File: backend/app/models/content.py
"""
Modèles pour le contenu du site vitrine - Section 2
Normalisé en 3FN
"""
from sqlalchemy import Column, String, Text, Integer, Boolean, Float, ForeignKey, JSON
from sqlalchemy.orm import relationship
from app.db.base_class import Base, BaseModel
import uuid


class News(Base, BaseModel):
    """
    Actualités et annonces - Section 2.1: Page d'accueil
    Déjà en 3FN — tous les attributs dépendent de la clé primaire.
    """
    __tablename__ = "news"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    title = Column(String, nullable=False)
    content = Column(Text, nullable=False)
    excerpt = Column(Text, nullable=True)  # Résumé court
    image_url = Column(String, nullable=True)
    author = Column(String, nullable=True)
    is_published = Column(Boolean, default=False, nullable=False)
    published_at = Column(String, nullable=True)  # ISO format datetime
    category = Column(String, nullable=True)  # Annonce, Actualité, etc.

    def __repr__(self):
        return f"<News {self.title}>"


class FAQ(Base, BaseModel):
    """
    Questions fréquemment posées - Section 2.3 et 2.4
    Déjà en 3FN.
    """
    __tablename__ = "faqs"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    question = Column(Text, nullable=False)
    answer = Column(Text, nullable=False)
    category = Column(String, nullable=True)  # Inscription, QCM, Sélection, etc.
    order = Column(Integer, default=0)  # Ordre d'affichage
    is_published = Column(Boolean, default=True, nullable=False)

    def __repr__(self):
        return f"<FAQ {self.question[:50]}>"


class Edition(Base, BaseModel):
    """
    Informations sur l'édition en cours - Section 2.3: Édition 2026
    Normalisée en 3FN : les données JSON sont éclatées en tables enfants.
    """
    __tablename__ = "editions"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    year = Column(Integer, unique=True, nullable=False)
    is_active = Column(Boolean, default=False, nullable=False)  # Une seule édition active

    # Présentation
    title = Column(String, nullable=False)
    description = Column(Text, nullable=True)

    # Relations normalisées (au lieu de colonnes JSON)
    timeline_phases = relationship("TimelinePhase", back_populates="edition", cascade="all, delete-orphan",
                                   order_by="TimelinePhase.phase_order")
    calendar_events = relationship("CalendarEvent", back_populates="edition", cascade="all, delete-orphan",
                                    order_by="CalendarEvent.event_date")
    selection_criteria = relationship("SelectionCriterion", back_populates="edition", cascade="all, delete-orphan",
                                      order_by="SelectionCriterion.stage_order")
    edition_partners = relationship("EditionPartner", back_populates="edition", cascade="all, delete-orphan")

    def __repr__(self):
        return f"<Edition {self.year}>"


class TimelinePhase(Base, BaseModel):
    """
    Phases de la timeline d'une édition (3FN)
    Remplace le champ JSON timeline_data dans Edition.
    """
    __tablename__ = "timeline_phases"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    edition_id = Column(String, ForeignKey("editions.id", ondelete="CASCADE"), nullable=False)

    phase_order = Column(Integer, nullable=False)  # Ordre de la phase (1-6)
    title = Column(String, nullable=False)  # Ex: "Phase 1 : Inscription"
    description = Column(Text, nullable=True)
    start_date = Column(String, nullable=True)  # ISO date
    end_date = Column(String, nullable=True)  # ISO date
    is_current = Column(Boolean, default=False, nullable=False)

    # Relation
    edition = relationship("Edition", back_populates="timeline_phases")

    def __repr__(self):
        return f"<TimelinePhase {self.title}>"


class CalendarEvent(Base, BaseModel):
    """
    Événements du calendrier d'une édition (3FN)
    Remplace le champ JSON calendar_data dans Edition.
    """
    __tablename__ = "calendar_events"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    edition_id = Column(String, ForeignKey("editions.id", ondelete="CASCADE"), nullable=False)

    title = Column(String, nullable=False)
    description = Column(Text, nullable=True)
    event_date = Column(String, nullable=False)  # ISO date
    event_type = Column(String, nullable=True)  # "deadline", "event", "announcement"

    # Relation
    edition = relationship("Edition", back_populates="calendar_events")

    def __repr__(self):
        return f"<CalendarEvent {self.title}>"


class SelectionCriterion(Base, BaseModel):
    """
    Critères de sélection par étape d'une édition (3FN)
    Remplace le champ JSON selection_criteria dans Edition.
    """
    __tablename__ = "selection_criteria"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    edition_id = Column(String, ForeignKey("editions.id", ondelete="CASCADE"), nullable=False)

    stage = Column(String, nullable=False)  # "qcm", "regional", "bootcamp", "final"
    stage_order = Column(Integer, nullable=False)
    criterion = Column(Text, nullable=False)  # Description du critère
    min_score = Column(Float, nullable=True)  # Score minimum si applicable

    # Relation
    edition = relationship("Edition", back_populates="selection_criteria")

    def __repr__(self):
        return f"<SelectionCriterion {self.stage}: {self.criterion[:50]}>"


class EditionPartner(Base, BaseModel):
    """
    Association entre une édition et ses partenaires (3FN)
    Table de jointure Many-to-Many avec attributs supplémentaires.
    Remplace le champ JSON partners_data dans Edition.
    """
    __tablename__ = "edition_partners"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    edition_id = Column(String, ForeignKey("editions.id", ondelete="CASCADE"), nullable=False)
    partner_id = Column(String, ForeignKey("partners.id", ondelete="CASCADE"), nullable=False)

    role = Column(String, nullable=True)  # "sponsor", "technique", "institutionnel"
    contribution = Column(Text, nullable=True)

    # Relations
    edition = relationship("Edition", back_populates="edition_partners")
    partner = relationship("Partner", back_populates="edition_partners")

    def __repr__(self):
        return f"<EditionPartner edition={self.edition_id} partner={self.partner_id}>"


class PastEdition(Base, BaseModel):
    """
    Bilan des éditions passées - Section 2.2
    Normalisée en 3FN : les données JSON sont éclatées en tables enfants.
    """
    __tablename__ = "past_editions"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    year = Column(Integer, unique=True, nullable=False)

    # Participation internationale
    host_country = Column(String, nullable=True)
    num_students = Column(Integer, nullable=True)

    # Relations normalisées (au lieu de colonnes JSON)
    past_timeline_phases = relationship("PastTimelinePhase", back_populates="past_edition",
                                         cascade="all, delete-orphan", order_by="PastTimelinePhase.phase_order")
    gallery_images = relationship("GalleryImage", back_populates="past_edition", cascade="all, delete-orphan")
    testimonials = relationship("Testimonial", back_populates="past_edition", cascade="all, delete-orphan")
    achievements = relationship("Achievement", back_populates="past_edition", cascade="all, delete-orphan")
    press_releases = relationship("PressRelease", back_populates="past_edition", cascade="all, delete-orphan")
    edition_stats = relationship("EditionStat", back_populates="past_edition", cascade="all, delete-orphan")

    def __repr__(self):
        return f"<PastEdition {self.year}>"


class PastTimelinePhase(Base, BaseModel):
    """
    Timeline du parcours d'une édition passée (3FN)
    Remplace le champ JSON timeline_data dans PastEdition.
    """
    __tablename__ = "past_timeline_phases"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    past_edition_id = Column(String, ForeignKey("past_editions.id", ondelete="CASCADE"), nullable=False)

    phase_order = Column(Integer, nullable=False)
    title = Column(String, nullable=False)
    description = Column(Text, nullable=True)
    date = Column(String, nullable=True)

    # Relation
    past_edition = relationship("PastEdition", back_populates="past_timeline_phases")

    def __repr__(self):
        return f"<PastTimelinePhase {self.title}>"


class GalleryImage(Base, BaseModel):
    """
    Photos de la galerie d'une édition passée (3FN)
    Remplace le champ JSON gallery_urls dans PastEdition.
    """
    __tablename__ = "gallery_images"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    past_edition_id = Column(String, ForeignKey("past_editions.id", ondelete="CASCADE"), nullable=False)

    image_url = Column(String, nullable=False)
    caption = Column(String, nullable=True)
    order = Column(Integer, default=0)

    # Relation
    past_edition = relationship("PastEdition", back_populates="gallery_images")

    def __repr__(self):
        return f"<GalleryImage {self.image_url}>"


class Testimonial(Base, BaseModel):
    """
    Témoignages d'une édition passée (3FN)
    Remplace le champ JSON testimonials_data dans PastEdition.
    """
    __tablename__ = "testimonials"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    past_edition_id = Column(String, ForeignKey("past_editions.id", ondelete="CASCADE"), nullable=False)

    author_name = Column(String, nullable=False)
    author_role = Column(String, nullable=True)  # "élève", "mentor", "parent"
    content = Column(Text, nullable=False)
    video_url = Column(String, nullable=True)
    photo_url = Column(String, nullable=True)

    # Relation
    past_edition = relationship("PastEdition", back_populates="testimonials")

    def __repr__(self):
        return f"<Testimonial by {self.author_name}>"


class Achievement(Base, BaseModel):
    """
    Distinctions et performances d'une édition passée (3FN)
    Remplace le champ JSON achievements_data dans PastEdition.
    """
    __tablename__ = "achievements"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    past_edition_id = Column(String, ForeignKey("past_editions.id", ondelete="CASCADE"), nullable=False)

    title = Column(String, nullable=False)
    description = Column(Text, nullable=True)
    category = Column(String, nullable=True)  # "médaille", "mention", "classement"
    rank = Column(Integer, nullable=True)

    # Relation
    past_edition = relationship("PastEdition", back_populates="achievements")

    def __repr__(self):
        return f"<Achievement {self.title}>"


class PressRelease(Base, BaseModel):
    """
    Communiqués de presse d'une édition passée (3FN)
    Remplace le champ JSON press_releases dans PastEdition.
    """
    __tablename__ = "press_releases"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    past_edition_id = Column(String, ForeignKey("past_editions.id", ondelete="CASCADE"), nullable=False)

    title = Column(String, nullable=False)
    source = Column(String, nullable=True)  # Nom du média
    url = Column(String, nullable=True)
    published_at = Column(String, nullable=True)  # ISO date

    # Relation
    past_edition = relationship("PastEdition", back_populates="press_releases")

    def __repr__(self):
        return f"<PressRelease {self.title}>"


class EditionStat(Base, BaseModel):
    """
    Statistiques comparatives d'une édition passée (3FN)
    Remplace le champ JSON stats_data dans PastEdition.
    """
    __tablename__ = "edition_stats"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    past_edition_id = Column(String, ForeignKey("past_editions.id", ondelete="CASCADE"), nullable=False)

    metric_name = Column(String, nullable=False)  # "nb_inscrits", "nb_filles", "score_moyen"
    metric_value = Column(Float, nullable=False)
    metric_unit = Column(String, nullable=True)  # "%", "candidats", "points"

    # Relation
    past_edition = relationship("PastEdition", back_populates="edition_stats")

    def __repr__(self):
        return f"<EditionStat {self.metric_name}={self.metric_value}>"


class Partner(Base, BaseModel):
    """
    Partenaires institutionnels - Section 2.4: Pages institutionnelles
    Déjà en 3FN. Ajout de la relation avec EditionPartner.
    """
    __tablename__ = "partners"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    name = Column(String, nullable=False)
    logo_url = Column(String, nullable=True)
    description = Column(Text, nullable=True)
    website_url = Column(String, nullable=True)
    order = Column(Integer, default=0)
    is_active = Column(Boolean, default=True, nullable=False)

    # Relation normalisée
    edition_partners = relationship("EditionPartner", back_populates="partner")

    def __repr__(self):
        return f"<Partner {self.name}>"


class Page(Base, BaseModel):
    """
    Pages institutionnelles - Section 2.4
    À propos, Mission, Contact, etc.
    Déjà en 3FN (meta_data est acceptable en JSON car c'est un
    bag de propriétés libres pour le SEO, sans structure fixe).
    """
    __tablename__ = "pages"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    slug = Column(String, unique=True, nullable=False)  # a-propos, mission, contact
    title = Column(String, nullable=False)
    content = Column(Text, nullable=True)  # HTML/Markdown
    meta_data = Column(JSON, nullable=True)  # SEO, images, etc.
    is_published = Column(Boolean, default=True, nullable=False)

    def __repr__(self):
        return f"<Page {self.slug}>"
